FEC-10923 Expose config to forceWidevineL3 on device

GouravSna opened the pull request:
### Description of the Changes

1. If the device codec is known to fail if security level L1 is used then set flag to true, it will force the player to use Widevine L3.
API will work only SDK level 18 or above.

`player.getSettings().forceWidevineL3Playback(true);` // Online Playback
`localAssetsManager.forceWidevineL3Playback(true);` // Offline Playback

2. ErrorCallback changes: Error in case of Output Protection Level (OPL), we are sending as `DRM_ERROR` to the apps with error message `DRM_ERROR:Required output protections are not active`

**NOTE:** For point#1 Application should keep a list of device where they are facing issue for playing the assets on L1 device. Setting this flag to all the device without knowing the side effects is not recommended. It can lead to fatal errors in playback.

giladna committed:
phase1 try to force L3 - no config yet

GouravSna committed:
- setForceWidevineL3Playback API

giladna commented on playkit/src/main/java/com/kaltura/playkit/drm/DeferredDrmSessionManager.java:
```diff
@@ -81,19 +85,21 @@ public void setMediaSource(PKMediaSource mediaSource) {
+            drmSessionManager = null;
+            return;
+        }
+        drmCallback.setLicenseUrl(license);
```
please add null protection

GouravSna replied:
Done

giladna commented on playkit/src/main/java/com/kaltura/playkit/drm/DeferredDrmSessionManager.java:
```diff
@@ -123,6 +129,33 @@ public DrmSession acquireSession(Looper playbackLooper, @Nullable EventDispatche
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2 && isForceWidevineL3Playback) {
+            drmSessionManager = new DefaultDrmSessionManager.Builder()
+                    //.setUuidAndExoMediaDrmProvider(MediaSupport.WIDEVINE_UUID, FrameworkMediaDrm.DEFAULT_PROVIDER) // Not using default provider
```
the if condition for L3 has to call .setUuidAndExoMediaDrmProvider in the new format 
else it has to be call in the previous way

                    .setUuidAndExoMediaDrmProvider(MediaSupport.WIDEVINE_UUID, FrameworkMediaDrm.DEFAULT_PROVIDER) // Not using default provider

GouravSna replied:
Done

GouravSna committed:
- Added null protection for drmCallback
- Added setUuidAndExoMediaDrmProvider with DEFAULT value if not forced to L3 playback
- Code refactor

GouravSna committed:
Merge branch 'dev' into FEC-10923

# Conflicts:
#	playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java

GouravSna committed:
- API name refactor

giladna commented on playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java:
```diff
@@ -358,18 +359,32 @@ private MediaItem buildExoMediaItem(PKMediaSourceConfig sourceConfig) {
 
+    private DeferredDrmSessionManager getDeferredDRMSessionManager() {
+        final DrmCallback drmCallback = new DrmCallback(getHttpDataSourceFactory(null), playerSettings.getLicenseRequestAdapter());
+        drmSessionManager = new DeferredDrmSessionManager(mainHandler, drmCallback, drmSessionListener, playerSettings.allowClearLead(), playerSettings.isForceWidevineL3Playback());
```
return new....

GouravSna replied:
Done

giladna commented on playkit/src/main/java/com/kaltura/playkit/drm/DeferredDrmSessionManager.java:
```diff
@@ -123,6 +142,32 @@ public DrmSession acquireSession(Looper playbackLooper, @Nullable EventDispatche
+                            uuid -> {
+                                try {
+                                    FrameworkMediaDrm frameworkMediaDrm = FrameworkMediaDrm.newInstance(MediaSupport.WIDEVINE_UUID);
+                                    frameworkMediaDrm.setPropertyString(SECURITY_LEVEL_PROPERTY, WIDEVINE_SECURITY_LEVEL_3);
```
you have to build it also for the previous provider I think in case of default
lets discuss it

GouravSna replied:
Done

GouravSna committed:
- Code refactor for DrmSessionManager Builder

GouravSna committed:
- Added `forceWidevineL3Playback` API for offline playback

GouravSna committed:
- Passing Crypto Exception with clear error message

GouravSna committed:
- Separated DRM_ERROR and RENDERER_ERROR

giladna commented on playkit/src/main/java/com/kaltura/playkit/drm/DeferredDrmSessionManager.java:
```diff
@@ -123,6 +138,35 @@ public DrmSession acquireSession(Looper playbackLooper, @Nullable EventDispatche
+                                frameworkMediaDrm.setPropertyString(SECURITY_LEVEL_PROPERTY, WIDEVINE_SECURITY_LEVEL_3);
+                                return frameworkMediaDrm;
+                            } catch (UnsupportedDrmException e) {
+                                throw new IllegalStateException(e);
```
@GouravSna 
we better set 
```
drmSessionManager = DrmSessionManager.getDummyDrmSessionManager();
```

instead no?

GouravSna replied:
drmSessionManager has already been created at this time because this try/catch is callback from ExoMediaDrm.Provider$acquireExoMediaDrm(UUID uuid).

GouravSna replied:
Fixed. Returning the default FrameworkMediaDrm

GouravSna committed:
- Added forceWidevineL3Playback flag for unregisterAsset and checkAssetStatus API

GouravSna committed:
- Handling UnsupportedDrmException
- Creating default FrameworkMediaDrm in exception

giladna committed:
error log
